---
title: "Bicluster validation functions"
author: Daniel He
output:
  html_document:
    toc: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(librarian)
librarian::shelf(tidyverse, dplyr, ggplot2, biomaRt, limma, edgeR, reshape, mosbi, RColorBrewer,cowplot)
source(paste0(getwd(),"/functions/mosbiConvert.R"))
```

Below are a list of tools to validate biclusters generated from gene expression data. 

# Data pre-processing

Loading in sample mesothelioma dataset of log-normalized RNA counts. 

```{r message=FALSE}
file_clinical <- paste0("http://linkedomics.org/data_download/TCGA-MESO/Human__",
                        "TCGA_MESO__MS__Clinical__Clinical__01_28_2016__BI__",
                        "Clinical__Firehose.tsi")
file_rnaseq <- paste0("http://linkedomics.org/data_download/TCGA-MESO/Human__",
                      "TCGA_MESO__UNC__RNAseq__HiSeq_RNA__01_28_2016__BI__Gene",
                      "__Firehose_RSEM_log2.cct.gz")

clinical <- readr::read_tsv(file_clinical)
rnaseq <- readr::read_tsv(file_rnaseq)

clinical[1:5, 1:3]
rnaseq[1:5,1:3]

# transposing clinical df so that rows corresponding to matrix columns
clinical <- clinical %>% column_to_rownames("attrib_name") %>% t() %>% as.data.frame() %>% rownames_to_column("ID")

# setting gene names to rownames
rnaseq <- rnaseq %>% column_to_rownames("attrib_name")
```

Checking if sample orders match for both dataframes:

```{r}
# testing if order of data matrix and metadata match
test_match_order <- function(x,y) {
  if (all(x==y)) print('Samples match in same order.')
  if (!all(x==y) && all(sort(x)==sort(y))) print('Samples match in wrong order, fix it.')
  if (!all(x==y) && !all(sort(x)==sort(y))) print('No match. Something is wrong.')
}
test_match_order(colnames(rnaseq),clinical$ID)

# fixing the order of rnaseq dataframe
rnaseq <- rnaseq[,match(clinical$ID, colnames(rnaseq))]
test_match_order(colnames(rnaseq),clinical$ID)
```

Checking if sample dataset look normalized:

```{r}
# Check distributions of samples using boxplots
boxplot(rnaseq, xlab="", ylab="Log2 counts per million",las=2)
# Let's add a blue horizontal line that corresponds to the median logCPM
title("Boxplots of logCPMs")
```

Applying z score transformation to data matrix

```{r}
z_score <- function(x, margin = 2) {
  z_fun <- function(y) {
    (y - mean(y, na.rm = TRUE)) / sd(y, na.rm = TRUE)
  }
  
  if (margin == 2) {
    return(apply(x, margin, z_fun))
  } else if (margin == 1) {
    return(t(apply(x, margin, z_fun)))
  }
}

# applying z score transformation
data_matrix = z_score(rnaseq, 1)

# checking if rows have SD of 1
transform(data_matrix, SD=apply(data_matrix,1, sd, na.rm = TRUE))[c(1:3,5000:5003),c(1:3, 88)]
```

# Biclustering

Running select biclustering algorithms with mosbi:

```{r}
set.seed(91)
if(!file.exists(paste0(getwd(),"/RDS/all_bics.RDS"))){
  # isa2
  BCisa <- mosbi::run_isa(data_matrix)
  # Plaid
  BCplaid <- mosbi::run_plaid(data_matrix)
  # QUBIC
  BCqubic <- mosbi::run_qubic(data_matrix, f=0.5)
  all_bics <- c(BCisa, BCplaid, BCqubic)
  saveRDS(all_bics, paste0(getwd(),"/RDS/all_bics.RDS"))
} else{
  all_bics <- readRDS(paste0(getwd(),"/RDS/all_bics.RDS"))
}

# filtering out overlapping biclusters using filter_subsets function from mosbi. Cannot find source code for this as it calls from a C++ function. I assume it uses simple overlap coefficient to determine similarities. 
all_bics <- filter_subsets(all_bics)
```

# Bicluster validation

## Calculating differential expression

Applying cluster validation function, which computes differential expression of bicluster vs non-bicluster samples and the level of overlap between bicluster genes and DE genes. Also computes signal-to-noise ratio (SNR) of the z-scored bicluster genes. This function takes the following arguments:

* **`biclusters`**: a list of biclusters from mosbi functions. Can also input a list of user-defined biclusters in the form of a list, with each element containing "row", "column", "rowname", "colname", and "algorithm". e.g. if isa2 identifies Gene 5, Gene 9, Gene 30, and Sample 2, Sample 8, Sample 33, Sample 87 as a bicluster, `row=c(5, 9, 30)`, `column=c(2, 8, 33, 87)`, `rowname=c("Gene 5", "Gene 9", "Gene 30")`, `colname=c("Sample 2", "Sample 8", "Sample 33", "Sample 87")`, and `algorithm="isa2"`. Function internally converts mosbi output to a list-type. 
* **`type`**: either "microarray" or "rnaseq". 
* **`eset`**: an expression set object of class `DGEList` containing raw counts. If unnormalized, will compute normalization factors using trimmed mean of M-values. Only used for `type="rnaseq"`. This does not need to be specified if a log-normalized matrix is provided with the `df` argument.
* **`df`**: a log-normalized matrix of gene expression values. Can be for `type="microarray"` or `type="rnaseq"`. 
* **`fdr`**: an integer that sets the false discovery rate threshold for differential expression. Defaults to 0.05.
* **`fold`**: an integer that sets a cutoff for log2 fold change. Defaults to 1.0.

```{r}
source(paste0(getwd(),"/functions/clusterValidation.R"))
if(!file.exists(paste0(getwd(),"/RDS/all_bics_validation.RDS"))){
  all_bics_validation <- clusterValidation(biclusters = all_bics, 
                                           type="rnaseq", 
                                           df = rnaseq, 
                                           fdr=0.05, fold=1.5)
  saveRDS(all_bics_validation, paste0(getwd(),"/RDS/all_bics_validation.RDS"))
} else {
  all_bics_validation <- readRDS(paste0(getwd(),"/RDS/all_bics_validation.RDS"))
}

str(all_bics_validation[[1]])
```

## Statistical validation of biclusters 

Computing statistical tests for outcomes of interest. For this example, we will examine the association of each bicluster with pathological stage of cancer. This function takes the following arguments:

* **`bicluster_list`**: a list of biclusters. Will convert a mosbi type object into a list type object.
* **`sample_info`**: a data frame containing metadata of interest for the statistical tests, with rows corresponding to samples. 
* **`sample_id`**: a character value indicating the column in sample_info containing sample names.
* **`outcome`**: a character value indicating the column in sample_info containing the outcome of interest.
* **`test`**: a character vector indicating the statistical tests to be computed. Can be any of `kruskal-wallis`, `chisq`, or `fisher`. `kruskal-wallis` requires the `order` argument to be specified. `fisher` will only work if there are two possible outcomes.
* **`order`**: if `kruskal-wallis` is specified as the statistical test, input a vector in ascending order of the outcome of interest. 

```{r warning=FALSE, message=FALSE}
source(paste0(getwd(),"/functions/bicluster_stat_test.R"))
all_bics_stats <- bicluster_stat_test(bicluster_list = all_bics_validation,
                    sample_info = clinical,
                    sample_id = "ID",
                    outcome = "pathologic_stage",
                    test = c("kruskal-wallis", "chisq"),
                    order = c("stagei", "stageii", "stageiii", "stageiv")
                    )
str(all_bics_stats[[1]])
```

# Bicluster visualization

Filtering biclusters with p<0.05, because none pass FDR<0.05. Also filtering out biclusters with fewer than 10 samples. 

```{r}
all_bics_stats_filtered <- Filter(function(x) all(x[["KW_pathologic_stage"]][["p.value"]] <= 0.05 &
                                                    length(x$column) >=10), all_bics_stats)
```

## Heatmap

Loading a function that plots heatmaps for a list of biclusters. It takes the arguments:

* **`bicluster_list`**: a list of biclusters. Will convert a mosbi type object into a list type object.
* **`eset`**: an expression set object of class `DGEList` containing raw counts. If unnormalized, will compute normalization factors using trimmed mean of M-values. Only used for `type="rnaseq"`. This does not need to be specified if a log-normalized matrix is provided with the `df` argument.
* **`df`**: a log-normalized matrix of gene expression values. Can be for `type="microarray"` or `type="rnaseq"`. 
* **`sample_info`**: a data frame containing metadata of interest for the statistical tests, with rows corresponding to samples. 
* **`sample_id`**: a character value indicating the column in sample_info containing sample names.
* **`values`**: a vector of column names in `sample_info` to be plotted as annotations on the heatmap.
* **`color_legend`**: a list of named character vectors corresponding to the elements in values. See below for example.
* **`gradient`**: a vector corresponding to the color gradient used for plotting.
* **`title_components`**: a vector of elements to be in the plot title. Options are to plot number of bicluster genes that are differentially expressed (`"DEG"`), overlap coefficient of bicluster and total DE genes (`"overlap"`), jaccard coefficient of bicluster and total DE genes (`"jaccard"`), signal-to-noise ratio of bicluster genes (`"SNR"`), Kruskal-Wallis p-value (`"K-W_p"`) and FDR (`"K-W_fdr"`) of biclusters, Chi-Squared p-value (`"chi_p"`) and FDR (`"chi_fdr"`), and Fisher p-value (`"fisher_p"`) and FDR (`"fisher_fdr"`) after running `bicluster_stat_test`. 
* **`multiple`**: an integer specifying how many columns to plot. This will remove the annotation legend from all heatmaps except for the heatmap in the last column. e.g. if `multiple = 3`, every 3rd heatmap will have an annotation legend.
* **`text_size`**: an integer specifying heatmap font size. Defaults to 5. 

```{r message=FALSE}
source(paste0(getwd(),"/functions/bicluster_to_heatmap.R"))

# creating a colour gradient
col_fun = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)

# creating an color legend with named vectors
ann_colors <- list(
  cluster = setNames(c("#999999","#009E73"), c("NonCluster", "BiCluster")),
  pathologic_stage = setNames(brewer.pal(9,"Pastel1")[1:length(levels(as.factor(clinical$pathologic_stage)))],
                              levels(as.factor(clinical$pathologic_stage))
                              ),
  histological_type = setNames(brewer.pal(8, "Accent")[1:length(levels(as.factor(clinical$histological_type)))],
                                levels(as.factor(clinical$histological_type))
                               ),
  radiation_therapy = setNames(brewer.pal(8, "Set2")[1:length(levels(as.factor(clinical$radiation_therapy)))],
                               levels(as.factor(clinical$radiation_therapy))
                               ),
  gender = setNames(brewer.pal(9,"Set1")[1:length(levels(as.factor(clinical$gender)))],
                    levels(as.factor(clinical$gender))
                    )
)

# generating heatmaps of biclusters. because there are 4 biclusters to be plotted, set multiple to 4 so that a grid plot can have ncol=4.
bicluster_plots <- bicluster_to_heatmap(bicluster_list = all_bics_stats_filtered,
                             df = rnaseq, 
                             sample_info = clinical, 
                             sample_ID = "ID", 
                             multiple = 4,
                             values = c("pathologic_stage", "histological_type", "radiation_therapy"),
                             color_legend = ann_colors,
                             gradient = col_fun,
                             title_components = c("DEG", "overlap", "jaccard", "SNR", "K-W_p","K-W_fdr"),
                             text_size=7)
```

Plotting the heatmaps using `cowplot`:

```{r fig.width=15, fig.height=6, out.width="100%"}
# in rel_widths, need to make the last column slightly larger to account for the annotation legend. Will require some manual adjustment.
plot_grid(plotlist=bicluster_plots, ncol=4,rel_widths=c(1,1,1,1.55))
```

